File: fannkuchredux_pmem.go
/*
* The Computer Language Benchmarks Game
* https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
*
* contributed by Oleg Mazurov, June 2010
* flag.Arg hack by Isaac Gouy
*
*/

package main

import (
	"fmt"
	"runtime"
	"flag"
	"strconv"
	"github.com/vmware/go-pmem-transaction/pmem"
	"github.com/vmware/go-pmem-transaction/transaction"
)

type Result struct {
	maxFlips int
	checkSum int
}

var (
	NCHUNKS = 720
	CHUNKSZ = 0
	NTASKS  = 0
	pmemFile = flag.String("file", "fkr_pmem", "fkr_pmem file name")
)
var n = 12
var Fact []int
var p_g [][]int
var pp_g [][]int
var count_g [][]int

func fannkuch( idxMin int, routine int, ch chan Result ) {

	idxMax := idxMin + CHUNKSZ
	if idxMax < Fact[n] {
		go fannkuch( idxMax, routine + 1, ch )
	} else {
		idxMax = Fact[n]
	}

	p     := p_g[routine]
    pp    := pp_g[routine]
	count := count_g[routine]

	// first permutation
	for i := 0; i<n; i++ {
		p[i] = i
	}
	for i, idx := n-1, idxMin; i>0; i-- {
		d := idx / Fact[i]
		count[i] = d
		idx = idx % Fact[i]

		copy( pp, p )
		for j := 0; j <= i; j++ {
		if j+d <= i {
				p[j] = pp[j+d]
		} else {
				p[j] = pp[j+d-i-1]
		}
		}
	}

	maxFlips := 1
	checkSum := 0

	for idx, sign := idxMin, true; ; sign = !sign {

		// count flips
		first := p[0]
	if first != 0 {
		flips := 1
		if p[first] != 0 {
		copy( pp, p )
		p0 := first
			for {
			flips++
			for i, j := 1, p0-1; i < j; i, j = i+1, j-1 {
				pp[i], pp[j] = pp[j], pp[i]
			}
			t := pp[p0]
			pp[p0] = p0
			p0 = t
			if pp[p0] == 0 {
				break
			}
			}
		}
		if maxFlips < flips {
		maxFlips = flips
		}
		if sign {
		checkSum += flips
		} else {
		checkSum -= flips
		}
	}

	if idx++; idx == idxMax {
		break
	}

	// next permutation
	if sign {
		p[0], p[1] = p[1], first
	} else {
		p[1], p[2] = p[2], p[1]
		for k := 2;; k++ {
			if count[k]++; count[k] <= k {
			break
		}
			count[k] = 0
		for j:=0; j<=k; j++ {
			p[j] = p[j+1]
		}
		p[k+1] = first
		first = p[0]
		}
	}
	}

	ch <- Result{ maxFlips, checkSum }
}

func printResult( n int, res int, chk int ) {
	fmt.Printf("%d\nPfannkuchen(%d) = %d\n", chk, n, res)
}

func validate_or_create(slice [][]int, name string) [][]int {
	arr := slice
	if arr == nil {
		println("array didnt exist")
		arr = pmem.Make(name, arr, NTASKS).([][]int)
	}

	if len(arr) != NTASKS {
		println("incorrect length")
		for i := range arr {
			pmem.Delete(name + strconv.Itoa(i))
		}
		pmem.Delete(name)
		arr = pmem.Make(name, arr, NTASKS).([][]int)
	}

	if arr[0] == nil {
		println("creating inner arrays")
		for i := range arr {
			arr[i] = pmem.Make(name + strconv.Itoa(i), arr[i], n).([]int)
		} 
	}

	if len(arr[0]) != n {
		for i := range arr {
			pmem.Delete(name + strconv.Itoa(i))
			arr[i] = pmem.Make(name + strconv.Itoa(i), arr[i], n).([]int)
		}
	}

	return arr
}

func main() {
	flag.Parse()

	if flag.NArg() > 0 { n,_ = strconv.Atoi( flag.Arg(0) ) }
	runtime.GOMAXPROCS(4)

	firstInit := pmem.Init(*pmemFile)
	if firstInit {
		Fact = pmem.Make("fact", Fact, n+1).([]int)
		txn("undo") {
			Fact[0] = 1
			for i := 1; i<len(Fact); i++ {
				Fact[i] = Fact[i-1] * i
			}
		}

		CHUNKSZ = (Fact[n] + NCHUNKS - 1) / NCHUNKS
		CHUNKSZ += CHUNKSZ%2
		NTASKS = (Fact[n] + CHUNKSZ - 1) / CHUNKSZ

		p_g = pmem.Make("p_g", p_g, NTASKS).([][]int)
		pp_g = pmem.Make("pp_g", pp_g, NTASKS).([][]int)
		count_g = pmem.Make("count_g", count_g, NTASKS).([][]int)

		txn("undo") {
			for i := range p_g {
				p_g[i] = pmem.Make("p_g" + strconv.Itoa(i), p_g[i], n).([]int)
				pp_g[i] = pmem.Make("pp_g" + strconv.Itoa(i), pp_g[i], n).([]int)
				count_g[i] = pmem.Make("count_g" + strconv.Itoa(i), count_g[i], n).([]int)
			}
		}
	} else {
		Fact = pmem.GetSlice("fact", Fact).([]int)

		if Fact == nil {
			Fact = pmem.Make("fact", Fact, n+1).([]int)
		}

		if len(Fact) != n+1 {
			pmem.Delete("fact")
			Fact = pmem.Make("fact", Fact, n+1).([]int)
		}

		if Fact[n] == 0 {
			txn("undo") {
				Fact[0] = 1
				for i := 1; i<len(Fact); i++ {
					Fact[i] = Fact[i-1] * i
				}
			}
		}

		CHUNKSZ = (Fact[n] + NCHUNKS - 1) / NCHUNKS
		CHUNKSZ += CHUNKSZ%2
		NTASKS = (Fact[n] + CHUNKSZ - 1) / CHUNKSZ
		
		p_g = pmem.GetSlice("p_g", p_g).([][]int)
		pp_g = pmem.GetSlice("pp_g", pp_g).([][]int)
		count_g = pmem.GetSlice("count_g", count_g).([][]int)

		txn("undo") {
			p_g = validate_or_create(p_g, "p_g")
			pp_g = validate_or_create(pp_g, "pp_g")
			count_g = validate_or_create(count_g, "count_g")
		}
	}
	ch := make(chan Result, NTASKS)

	go fannkuch(0, 0, ch)
	
	res := 0
	chk := 0
	for i := 0; i<NTASKS; i++ {
		r := <-ch
		if res < r.maxFlips {
			res = r.maxFlips
		}
		chk += r.checkSum
	}

	printResult( n, res, chk )
}

File: fannkuchredux_pmem.cpp
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// Contributed by Dave Compton
// Based on "fannkuch-redux C gcc #5", contributed by Jeremy Zerfas
// which in turn was based on the Ada program by Jonathan Parker and 
// Georg Bauhaus which in turn was based on code by Dave Fladebo, 
// Eckehard Berns, Heiner Marxen, Hongwei Xi, and The Anh Tran and 
// also the Java program by Oleg Mazurov.

#include <iostream>
#include <fstream>
#include <algorithm>
#include <math.h>
#include <pthread.h>
#include <libpmemobj/atomic_base.h>
#include <libpmemobj++/make_persistent.hpp>
#include <libpmemobj++/make_persistent_atomic.hpp>
#include <libpmemobj++/transaction.hpp>
#include <libpmemobj++/utils.hpp>
#include "pvector.hpp"
#define CREATE_MODE_RW (S_IWUSR | S_IRUSR)


using namespace std;
using namespace pmem::obj;

const char *LAYOUT = "fkrcplus";


class Permutation 
{
  public:
    // Permutation(int n, int64_t start);
    ~Permutation();
    void initialize(int n, int64_t start);
    void advance();
    int64_t countFlips() const;

  private:
     persistent_ptr<PVector <int>> count;
     persistent_ptr<PVector <int8_t>> current;
     size_t len;

};

struct root {
   persistent_ptr<PVector<Permutation>> perms;
   persistent_ptr<int64_t[]> fact;
};

pmem::obj::pool<root> pop;
persistent_ptr<int64_t[]> fact;

void initializeFact(int n)
{
    fact[0] = 1;
    for (auto i = 1; i <= n; ++i)
        fact[i] = i * fact[i - 1];
}

Permutation::~Permutation() {
    auto pool = pool_by_vptr(this);
    transaction::run(pool, [this] {
        delete_persistent<PVector<int>>(count);
        delete_persistent<PVector<int8_t>>(current);
    });
}

// 
// Initialize the current value of a permutation
// and the cycle count values used to advance .
// 
void Permutation::initialize(int n, int64_t start)
{
    if (count == NULL) {
        auto pool = pool_by_vptr(this);
        transaction::run(pool, [this, n] {
            count = make_persistent<PVector<int>>(n);
            current = make_persistent<PVector<int8_t>>(n);
        });
    } else {
        count->resize(n);
        current->resize(n);
    }

    len = n;

    // Initialize count 
    for (auto i = n - 1; i >= 0; --i) 
    {
        auto d = start / fact[i];
        start = start % fact[i];
        count->get(i) = d;
    }

    // Initialize current.
    for (auto i = 0; i < n; ++i)
        current->get(i) = i;

    for (auto i = n - 1; i >= 0; --i) 
    {
        auto d = count->get(i);
        auto b = current->begin();
        rotate(b, b + d, b + i + 1);
    }
}

//
// Advance the current permutation to the next in sequence.
// 
void Permutation::advance()
{
    for (auto i = 1; ;++i) 
    {
        // Tried using std::rotate here but that was slower.
        auto first = current->get(0);
        for (auto j = 0; j < i; ++j)
            current->get(j) = current->get(j + 1);
        current->get(i) = first;

        ++(count->get(i));
        if (count->get(i) <= i)
            break;
        count->get(i) = 0;
    }
}

//
// Count the flips required to flip 0 to the front of the vector.
//
// Other than minor cosmetic changes, the following routine is
// basically lifted from "fannkuch-redux C gcc #5"
//
inline int64_t Permutation::countFlips() const
{
    const auto n = current->cap();
    auto flips = 0;
    auto first = current->get(0);
    if (first > 0) 
    {
        flips = 1;

        int8_t temp[n];
        // Make a copy of current to work on. 
        for (size_t i = 0; i < n; ++i)
            temp[i] = current->get(i);


        // Flip temp until the element at the first index is 0
        for (; temp[first] > 0; ++flips) 
        {
            // Record the newFirst and restore the old
            // first at its new flipped position.
            const int8_t newFirst = temp[first];
            temp[first] = first;

            if (first > 2) 
            {
                int64_t low = 1, high = first - 1;
                do 
                {
                    swap(temp[low], temp[high]);
                    if (!(low + 3 <= high && low < 16))
                        break;
                    ++low;
                    --high;
                } while (1);
            }
            // Update first to newFirst that we recorded earlier.
            first = newFirst;
        }
    }
    return flips;
}

bool file_exists(const std::string &path) {
   std::ifstream f(path.c_str());
   return f.good();
}

int main(int argc, char **argv)
{

    std::string path = "fkr_cpp";
    const auto n = atoi(argv[1]);


    if (file_exists(path)) {
        std::cout << "opening existing pmem file\n";
        pop = pmem::obj::pool<root>::open(path, LAYOUT);
    } else {
        std::cout << "opening new pmem file\n";
        pop = pmem::obj::pool<root>::create(path, LAYOUT, 1024*1024*64, CREATE_MODE_RW);
    }

    auto root = pop.root();

    transaction::run(pop, [&root, n] {
        root->fact = make_persistent<int64_t[]>(32);
        fact = root->fact;
        initializeFact(n);
        root->perms = make_persistent<PVector<Permutation>>();
    });

    // Compute some factorials for later use.

    // blockCount works best if it is set to a multiple of the number
    // of CPUs so that the same number of blocks gets distributed to
    // each cpu.  The computer used for development (Intel i7-4700MQ)
    // had 8 "CPU"s (4 cores with hyperthreading) so 8, 16 and 24 
    // all worked well.


    auto blockCount = 24;
    if (blockCount > fact[n])
        blockCount = 1;
    const int64_t blockLength = fact[n] / blockCount;
    cout << "waddup1" << endl;

    root->perms->resize(blockCount);

    cout << "waddup" << endl;


    int64_t maxFlips = 0, checksum = 0;

    // Iterate over each block.
    #pragma omp parallel for \
        reduction(max:maxFlips) \
        reduction(+:checksum)

    for (int64_t blockStart = 0;
         blockStart < fact[n]; 
         blockStart += blockLength) 
    {
        auto perm = ceil((float) blockStart / blockLength);
        // first permutation for this block.
        root->perms->get(perm).initialize(n, blockStart);

        // Iterate over each permutation in the block.
        auto index = blockStart;
        while (1) 
        {
            const auto flips = root->perms->get(perm).countFlips();

            if (flips) 
            {
                if (index % 2 == 0)
                    checksum += flips;
                else
                    checksum -= flips;

                if (flips > maxFlips)
                    maxFlips = flips;
            }

            if (++index == blockStart + blockLength)
                break;

            // next permutation for this block.
            root->perms->get(perm).advance();
        }
    }

    // Output the results to stdout.
    cout << checksum << endl;
    cout << "Pfannkuchen(" << n << ") = " << maxFlips << endl;

    transaction::run(pop, [&root] {
        delete_persistent<int64_t[]>(root->fact, 32);
        delete_persistent<PVector<Permutation>>(root->perms);
    });
    pop.close();

    return 0;
}
